trieste/space.py

```diff
@@ -331,3 +331,140 @@ def __mul__(self, other: Box) -> Box:
 
     def __deepcopy__(self, memo: dict[int, object]) -> Box:
         return self
+
+
+class OrdinalSearchSpace(Box):
+    r"""
+    Ordinal search space representing a discrete :math:`D`-dimensional box in
+    :math:`\mathbb{R}^D`. This class inherit :class:`Box` and generates evenly-spaced value
+    in an interval for each dimension. For example randomly sample from this search space
+    will give:
+
+        >>> tf.random.set_seed(1945)
+        >>> ordinalsp = OrdinalSearchSpace([-2.0, -2.0], [2.0, 2.0], [0.2, 0.3])
+        >>> sample = ordinalsp.sample(5)
+        >>> sample.numpy()
+        array([[ 0. , -0.6],
+               [ 0.6, -0.6],
+               [-1.2,  1.2],
+               [-0.4,  1.5],
+               [ 1.2,  0. ]])
+
+    """
+
+    def __init__(
+        self,
+        lower: Sequence[float] | TensorType,
+        upper: Sequence[float] | TensorType,
+        stepsizes: Sequence[float] | TensorType,
+    ):
+
+        r"""
+        If ``lower``, ``upper`` and ``stepsize`` are `Sequence`\ s of floats
+        (such as lists or tuples), they will be converted to tensors of dtype `tf.float64`.
+
+        :param lower: The lower (inclusive) bounds of the Ordinal Space. Must have shape [D]
+            for positive D, and if a tensor, must have float type.
+        :param upper: The upper (inclusive) bounds of the Ordinal Space. Must have shape [D]
+            for positive D, and if a tensor, must have float type.
+        :param stepsizes: The step size of each dimension of the Ordinal Space. Must have
+            shape [D] for positive D, and if a tensor, must have float type.
+        :raise ValueError (or InvalidArgumentError): If any of the following are true:
+
+            - ``lower``, ``upper`` and ``stepsizes`` have invalid shapes.
+            - ``lower``, ``upper`` and ``stepsizes`` do not have the same floating point type.
+            - ``upper`` is not greater than ``lower`` across all dimensions.
+        """
+
+        super().__init__(lower, upper)
+        tf.debugging.assert_shapes([(lower, ["D"]), (stepsizes, ["D"])])
+        tf.assert_rank(stepsizes, 1)
```

Comment:
Should we check that the stepsize values aren't 0 or negative?

Reply:
Make sense, I'll add it.